### ISSUE TITLE
Crash when accessing/writing to compound variable

### ISSUE THREAD

**Issue by equaeghe** [Unidata/netcdf4-python#705]
Versions:
* netcdf4-python-1.2.2
* netcdf-4.4.1.1
* hdf5-1.8.18

I have not been able to get past crashes trying to access and write to a compound variable.
The following minimal example worked out in the IDE Spyder illustrates what I am seeing:

```python
    import netCDF4 as nc
    import numpy as np

    f = nc.Dataset('test.nc', 'w')

    val_np = np.dtype([("val", 'f4'), ("acc", 'i1')])
    val = f.createCompoundType(val_np, "value with accuracy")
    flags = {"missing": -1, "raw": 0, "doubtful quality": 1, "quality controlled": 2}
    flag = f.createEnumType('i1', "flag", flags)
    avg_np = np.dtype([("avg", val_np), ("flag", flag)])
    avg = f.createCompoundType(avg_np, "avg_statistic")

    f.createDimension("time", 3)
    # OUT: <class 'netCDF4._netCDF4.Dimension'>: name = 'time', size = 3
    f.createDimension("level", 2)
    # OUT: <class 'netCDF4._netCDF4.Dimension'>: name = 'level', size = 2

    f.createVariable("name", avg, ("time", "level"), zlib=True, complevel=9, fill_value=False)
    # OUT:
    # <class 'netCDF4._netCDF4.Variable'>
    # compound name(time, level)
    # compound data type: [('avg', [('val', '<f4'), ('acc', 'i1')]), ('flag', 'i1')]
    # unlimited dimensions: 
    # current shape = (3, 2)

    f["name"]
    # OUT: 
    # <class 'netCDF4._netCDF4.Variable'>
    # compound name(time, level)
    # compound data type: [('avg', [('val', '<f4'), ('acc', 'i1')]), ('flag', 'i1')]
    # unlimited dimensions: 
    # current shape = (3, 2)

    f["name"][:]
    # OUT: 
    # array([[((0.0, 0), 0), ((0.0, 0), 0)],
    #       [((0.0, 0), 0), ((0.0, 0), 0)],
    #       [((0.0, 0), 0), ((0.0, 0), 0)]], 
    #      dtype=[('avg', [('val', '<f4'), ('acc', 'i1')]), ('flag', 'i1')])

    f["name"][0,0] = np.array([((1.1, 2), 3)], avg_np)

    f["name"][0,0]
    # OUT: ((1.100000023841858, 2), 3)

    f["name"][0,:]
    # OUT: 
    # array([((1.100000023841858, 2), 0), ((-2.522968227147279e-32, -103), 99)], 
    #      dtype=[('avg', [('val', '<f4'), ('acc', 'i1')]), ('flag', 'i1')])

    f["name"][:]

    Kernel died, restarting
```
When the crash happens actually varies, sometimes it is when I first try to access the variable, sometimes it is later than the sequence shown above. This makes it quite strange. (Perhaps it is memory-allocation related and it depends on what junk data is at memory locations the library tries to access. I say this because of the random values appearing in the second array element.)

**Comment by jswhit** [comment]
Confirmed.  On my machine, moving the assignment

     f["name"][0,0] = np.array([((1.1, 2), 3)], avg_np)

before the variable is accessed, i.e. before the line

    f["name"][:]

gets rid of the crash.  Not sure whether this is a c library issue or a python interface issue - I'll try to create a C version of your test.

**Comment by jswhit** [comment]
The crash seems to occur whenever a compound variable is accessed before any data is written to it.  Here's a barebones example:

```python
import netCDF4 as nc
import numpy as np
f = nc.Dataset('test.nc', 'w')
val_np = np.dtype([("val", 'f4'), ("acc", 'i1')])
val = f.createCompoundType(val_np, "value with accuracy")
f.createDimension("time", 3)
f.createVariable("name", val, "time")
print f["name"][:]
f.close()
```

Commenting out the variable access, or closing and reopening the file before the variable access makes the crash go away.

**Comment by equaeghe** [comment]
Hmm, I haven't been able to get that barebones example to crash yet. However, assignment is showing issues:
```python
import netCDF4 as nc
import numpy as np
f = nc.Dataset('test.nc', 'w')
val_np = np.dtype([("val", 'f4'), ("acc", 'i1')])
val = f.createCompoundType(val_np, "value with accuracy")
f.createDimension("time", 3)
f.createVariable("name", val, "time")

Out[1]: 
<class 'netCDF4._netCDF4.Variable'>
compound name(time)
compound data type: [('val', '<f4'), ('acc', 'i1')]
unlimited dimensions: 
current shape = (3,)

f["name"][:] = np.array([(1.2, 3), (1.3, 4), (1.4, 5)], dtype=val_np)

f["name"][:]
Out[3]: 
array([(1.2000000476837158, 3), (1.5111572745182865e+23, 102),
       (3.901747418096734e-40, 0)], 
      dtype=[('val', '<f4'), ('acc', 'i1')])
```
Also, in my full code from which I derived my initial example in the issue, I never do access before assignment and do get crashes. But assuming these things may be related, we shouldn't focus on that yet.

(There is an extra complication in my full code, that I try to write to one field of the compound datatype, e.g.,
```python
f["name"][:]['val'] = 1.1
```
which doesn't seem to work when tried on our barebones example, but that may be a different issue—perhaps I shouldn't expect that to work.)

**Comment by jswhit** [comment]
It looks like crashes are not always repeatable. 

Assigning to one field of a compound type is not supported by netcdf.

**Comment by jswhit** [comment]
Can you try adding a `sync` to your code before you try to access any data in the netcdf variable?

**Comment by equaeghe** [comment]
> Assigning to one field of a compound type is not supported by netcdf.

Should I separately feature-request triggering a Python exception when trying?

> Can you try adding a `sync` to your code before you try to access any data in the netcdf variable?

I will be updating my full code based on the things we (you) have discovered and report back, to confirm or not that `sync`ing can be used as a workaround. However, this won't be right away. In any case, the issues exhibited by the barebones example should be fixed anyway, to give us some more confidence in using compound types.

**Comment by jswhit** [comment]
Regarding accessing the variable by field - it would be nice to warn to trigger an execption but I don't know how we would catch that.

Of course I want to fix the segfault - I'm just trying to gather data on the underlying cause.  So far I have not been able to trigger the crash in a simple C program.

**Comment by equaeghe** [comment]
> Can you try adding a `sync` to your code before you try to access any data in the netcdf variable?

Syncing does not seem to be sufficient:
```python
f['PTB/Ld']
Out[2]: 
<class 'netCDF4._netCDF4.Variable'>
compound Ld(time, level_PTB)
    units: hPa
compound data type: [('avg', [('val', '<f4'), ('acc', 'i1')]), ('flag', 'i1')]
path = /PTB
unlimited dimensions: 
current shape = (683856, 2)

f.sync()

f['PTB/Ld']
Out[4]: 
<class 'netCDF4._netCDF4.Variable'>
compound Ld(time, level_PTB)
    units: hPa
compound data type: [('avg', [('val', '<f4'), ('acc', 'i1')]), ('flag', 'i1')]
path = /PTB
unlimited dimensions: 
current shape = (683856, 2)

f['PTB/Ld'][:]

Kernel died, restarting
```

However, closing and reopening does seem to provide a workaround:
```python
f['PTB/Ld']
Out[2]: 
<class 'netCDF4._netCDF4.Variable'>
compound Ld(time, level_PTB)
    units: hPa
compound data type: [('avg', [('val', '<f4'), ('acc', 'i1')]), ('flag', 'i1')]
path = /PTB
unlimited dimensions: 
current shape = (683856, 2)

f.close()

f = nc.Dataset('Fino1.nc', 'r')

f['PTB/Ld']
Out[6]: 
<class 'netCDF4._netCDF4.Variable'>
compound Ld(time, level_PTB)
    units: hPa
compound data type: {'names':['avg','flag'], 'formats':[{'names':['val','acc'], 'formats':['<f4','i1'], 'offsets':[0,4], 'itemsize':8},'i1'], 'offsets':[0,8], 'itemsize':12, 'aligned':True}
path = /PTB
unlimited dimensions: 
current shape = (683856, 2)

f['PTB/Ld'][:]
Out[7]: 
array([[((nan, 127), -1), ((nan, 127), -1)],
       [((nan, 127), -1), ((nan, 127), -1)],
       [((nan, 127), -1), ((nan, 127), -1)],
       ..., 
       [((nan, 127), -1), ((nan, 127), -1)],
       [((nan, 127), -1), ((nan, 127), -1)],
       [((nan, 127), -1), ((nan, 127), -1)]], 
      dtype={'names':['avg','flag'], 'formats':[{'names':['val','acc'], 'formats':['<f4','i1'], 'offsets':[0,4], 'itemsize':8},'i1'], 'offsets':[0,8], 'itemsize':12, 'aligned':True})
```

Notice how the value for `compound data type` changes. Perhaps the specification of the compound data type is not correctly recorded in the `Variable` until only after a full close. I would assume that a sync should already do this at least or even that this should be done immediately, as it may be the root issue.

**Comment by jswhit** [comment]
OK, I think I've found the problem.  The numpy datatype has to have the `align=True` so padding will be added to the fields to match what a C compiler would output for a similar C-struct.   I attempt to do this in the python interface by re-defining the numpy datatype with

    dt = numpy.dtype(dt,align=True)

However, this apparently only works if `dt`is specified as a dictionary, not a list (see https://docs.scipy.org/doc/numpy-1.10.0/reference/generated/numpy.dtype.html).  

It works when you close and re-open the file since the numpy datatype is rebuilt as a dictionary with the `aligned` key set to True.  That's why the datatype in your output looks different before and after the file is closed and reopened.

Not sure what the best way to fix this is yet, but a workaround is to specify the numpy data type in `createCompoundType` as a dictionary.

**Comment by equaeghe** [comment]
> Potential fix in pull request #706
I see: you require the compound datatype to be defined through a `dict` and set `align=True` if this hasn't been set. I guess it is not possible to transform a `list`-based dtype to a `dict`-based one? (I'll have a look at this.)

**Comment by jswhit** [comment]
Some of the tests are failing with pull request #706.  I'm beginning to think the only reliable way to fix this is to require aligned datatypes in createCompoundType.

**Comment by equaeghe** [comment]
I looked at whether one could transform a (any?) `dtype` to an aligned one. I'm using Python 3.4 and so don't know about 2.7, but it seems to be possible:

```python
import numpy as np

dt = np.dtype([('a', 'f4'),('b', 'i1')])

dt.fields
Out: mappingproxy({'a': (dtype('float32'), 0), 'b': (dtype('int8'), 4)})

dt.isalignedstruct
Out: False

dt.fields.keys()
Out: dict_keys(['a', 'b'])

dta = np.dtype({'names': list(dt.fields.keys()), 'formats': [v[0]for v in dt.fields.values()]}, align=True)

dta
Out: dtype({'names':['a','b'], 'formats':['<f4','i1'], 'offsets':[0,4], 'itemsize':8}, align=True)
```

If this generalizes, it might be used as a preprocessing step in `createCompoundDatatype`.

**Comment by jswhit** [comment]
Yes, I've implemented something very similar.  It gets a little tricky with nested structured data types.  One thing I'm still not sure of is whether casting the data to the new aligned data type works for nested  types.

**Comment by jswhit** [comment]
@equaeghe, call you try pull request #706 with your original code and report back here?

**Comment by equaeghe** [comment]
> @equaeghe, call you try pull request #706 with your original code and report back here?

Hmm, I'm afraid that this code is not in version control; but anyway, the original code also contained lots of ‘assignments to compound fields’, so it wouldn't be a good test. My current code can be made into a test by setting `align=False` instead of `align=True`.

However, a bigger issue is installing the development code, as I'm using my distribution's (Gentoo's) packaging system. I guess I clone the repository locally, checkout branch ‘issue705’ and install it? (I can try, but this may take a while for me to find the time, as I don't do this kind of thing regularly.)

**Comment by jswhit** [comment]
Yes, you can clone the repo, switch to the issue705 branch and then install it locally.

**Comment by equaeghe** [comment]
To set a baseline, with versions:

    netcdf4-python-1.2.2
    netcdf-4.4.1.1
    hdf5-1.8.18

I get with my current code:

1. `align=True`, no `close` (or `sync`) until after all data is written: nc file of 69 MiB without known problems
2. `align=False`, no `close` (or `sync`) until after all data is written: nc file of 48 MiB without known problems
3. `align=True`, `close` between initializing variables and filling with real data: computation has to be aborted, because it keeps on running at some point (not yet clear why)
4. `align=False`, `close` between initializing variables and filling with real data: nc file created, but where most variable components contain nonsense values

The content of (1) and (2) appears to be the same; I checked this by going over all variables in those files and comparing them for equality. Also the compound types' definition seems to be the same after closing and reopening. I assume compression is different between the two for some reason.

I guess (4) contains nonsense because the (unaligned) dtype definitions used for the data that is written to the files is different from the (aligned) dtype established when closing.

My next step is to try and install the issue705 branch code and try the same four cases. (However, my code does contain nested compound types, so if I understand the patch correctly, that'll block any further investigation; I'll just try 1.2.29 then, to see what that does.)

**Comment by jswhit** [comment]
Pull request #706 should work with nested types.

**Comment by equaeghe** [comment]
Now, with versions:

```
netcdf4-python-1.3.0 (issue705 branch)
netcdf-4.4.1.1
hdf5-1.8.18
```

I get with my current code:

1. `align=True`, no `close` (or `sync`) until after all data is written: nc file of 69.2 MiB without known problems
2. `align=False`, no `close` (or `sync`) until after all data is written: nc file of 62.6 MiB without known problems
3. `align=True`, close between initializing variables and filling with real data: computation has to be aborted, because it keeps on running at some point (not yet clear why)
4. `align=False`, close between initializing variables and filling with real data: computation has to be aborted, because it keeps on running at some point (not yet clear why)

So with the new version, (3) and (4) seem to coincide qua (bad) behavior. Again, the content of (1) and (2) appears to be the same; also the same as (1) and (2) from the baseline.

So the new code does avoid creating variables filled with nonsense. Also, the segfaults are gone. However, code that should run, i.e., (3) and (4), does not and there are still size differences that are unexplained.

**Comment by jswhit** [comment]
What do you mean by 'size differences that are unexplained'?  Could you post a script that triggers the behavior you see in #3 and #4?

**Comment by jswhit** [comment]
There are currently restrictions on how complex the nested structured types can be.  For example

```python
import numpy as np
import netCDF4
dtype1 = np.dtype({'names':['i','j'],'formats':['<i2','<i8']},align=True)
dtype2 = np.dtype({'names':['x','y'],'formats':['<f4',('<f8', (3, 2))]},align=True)
dtype3 = np.dtype({'names':['xx','yy'],'formats':[(dtype1,(2,2)),dtype2]},align=True)
# the only below works fine (removing the (2,2) shape parameter)
#dtype3 = np.dtype({'names':['xx','yy'],'formats':[dtype1,dtype2]},align=True)
print netCD4. _set_alignment(dtype3)
```

will raise an error 

    TypeError: nested structured dtype arrays not supported

This can be fixed eventually, but for right now it is disallowed.

**Comment by equaeghe** [comment]
> What do you mean by 'size differences that are unexplained'?

Well, given that the content of the file is the same and the definition of variables as well, I would expect the files to have the same size. Now, just because of the difference in `align` setting, there is a size difference (69 vs 62 MiB).

> Could you post a script that triggers the behavior you see in 3 and 4?

I'll see if I can cook something up. However, my impression is that the actual data is involved, as it is always during the filling of the same variable that the problem (computation not stopping) occurs. Given that I am not allowed to share the actual data, this makes it more difficult. I do not think this should hold back this patch, which should already make working with compound data structures more reliable.

**Comment by jswhit** [comment]
I don't understand the file size issue.  The align flag is always set in the definition of the datatype, so the padding is added before the data is written to the file.  You can't write unaligned data through the netcdf4-python interface - if you pass it unaligned data it will be cast to the corresponding aligned data type before it is written to the file.  When you read the data from the file, it will always return a numpy structured array with the `isalignedstruct` flag set to True.

The second issue you mention ("computation not stopping") also is a mystery to me.  It should not matter whether you fill the variable after the file is closed and reopened, or just after you create the variable (before the file is closed).  I've tried some simple programs that do it both ways, and have not seen any issues.

**Comment by jswhit** [comment]
Perhaps the issue is that your structured (compound type) netcdf data is being provided by another application (not written through the python interface) so it contains structures that are not aligned?  If so there will be a difference in file size, because when you re-write the data from the python interface it will be aligned (and have extra padding, making the file size larger).

**Comment by equaeghe** [comment]
> Perhaps the issue is that your structured (compound type) netcdf data is being provided by another application (not written through the python interface) so it contains structures that are not aligned?

No, everything passes through netcdf4-Python, which is fed numpy arrays.

**Comment by jswhit** [comment]
There must be some differences in the memory layout of the compound types that are being written out in your tests 1 and 2.  Perhaps the members are oriented differently in the structure, so the offsets are different?  Have you checked to see that numpy arrays are actually equivalent when you read them in from the two files?

**Comment by equaeghe** [comment]
> There must be some differences in the memory layout of the compound types that are being written out in your tests 1 and 2. Perhaps the members are oriented differently in the structure, so the offsets are different?

I did not yet check explicitly for that, although I did check some of those manually, where they turned out to be identical. What I did not check is chunk sizes, which would influence compression.

> Have you checked to see that numpy arrays are actually equivalent when you read them in from the two files?

Yes, I used the following code:
```python
for g in fn.groups:
    for d in fn[g].variables:
        a = fa[g][d][:]
        n = fn[g][d][:]
        c =  ~(a == n)
        if n.dtype.names is not None:
            for s in {'min', 'max', 'avg', 'std'}:
                if s in n.dtype.names:
                    c = c & ~(np.isnan(n[s]['val']) & np.isnan(a[s]['val']))
            if 'val' in n.dtype.names:
                c = c & ~(np.isnan(n['val']) & np.isnan(a['val']))
        print(g, d, np.any(c))
```
where `fa` and `fn` are the two nc files with verified identical structure and the compound datatypes are of the form `[('avg', [('val', 'f4'), ('acc', 'i1')]), (…)]` or `[('val', 'f4'), (…)]`.

**Comment by jswhit** [comment]
OK - that's good to know. I'm not too worried about the change in size of the files then.  That leaves only your problem with the program hanging when you are writing data.  It would be nice to have a test case that triggers that.

**Comment by jswhit** [comment]
I'm going to go ahead and merge pull request #706.  Please open a new issue for the other problem you mentioned if it persists.

**Comment by equaeghe** [comment]
> I don't understand the file size issue. The align flag is always set in the definition of the datatype, so the padding is added before the data is written to the file.

For the size issue, the alignment isn't the problem. It seems to be the padding. Data in `('f4', 'i1')` compound arrays compresses worse than the same data in `('f4', 'i4')` arrays. I guess this is because the `'i1'` part is padded to 4 bytes, but the content of those 3 other bytes is not constant (0 or 1) but whatever random junk was in memory at that time. I guess netCDF or netCDF4-python should make sure the padding is constant (0 or 1). The difference is substantial; 20% for my real-life dataset.

**Comment by jswhit** [comment]
Interesting - the padding is handled by numpy and I don't know of a way to control how the memory is initialized in the padding. I will investigate.

**Comment by jswhit** [comment]
I don't see any way around the increased size when the structures are padded - the extra bytes will result in a larger file.   There's no way to control how the structures are padded to make the compression more efficient.

**Comment by equaeghe** [comment]
> the padding is handled by numpy and I don't know of a way to control how the memory is initialized in the padding.

What function or functionality of numpy is being used? I may file an issue to the numpy tracker, given that does have a substantial impact (non-byte-for-byte identical files, memory contents being exposed, etc.).

**Comment by jswhit** [comment]
align=True when creating the dtype

**Comment by equaeghe** [comment]
My experiments so far indicate that numpy pads with zeros:

```python
>>> import numpy as np

>>> c = np.dtype([('f4', 'f4'), ('i1', 'i1')], align=True)
>>> c
dtype({'names':['f4','i1'], 'formats':['<f4','i1'], 'offsets':[0,4], 'itemsize':8}, align=True)

>>> a = np.array([(0, 0)], dtype=c)
>>> a.view('i8')[0]
0
```

**Comment by jswhit** [comment]
Then maybe it's just the extra space the padding takes that accounts for the file size differences?

**Comment by equaeghe** [comment]
> Then maybe it's just the extra space the padding takes that accounts for the file size differences?

No, for my dataset, if I store an integer flag within a compound dataset as `'i1'`, which gets padded to 4 bytes, it takes *substantially more* space than if I store it as `'i4'`, also taking up 4 bytes. So the uncompressed sizes are the same, but the compression is better when using the data type that does not need to be padded. That is why I thought that the padding was random, because for `'i4'`, there is implicit padding with zeros (or ones, as the flags could take the values -1, 0, 1, 2). (Actually the flags are defined as an enum, but that shouldn't matter, I think.)
